### ISSUE TITLE
How should we add extension points to the protocol?

### ISSUE THREAD

**Issue by bytemaster** [cryptonomex/graphene#141]
Global Properties and Fees are just two examples of places where we may want to add values in the future.  

We have several different techniques in use already:
block_header a vector<header_extension> which includes a static_variant of possible extension types.
fee_schedule has a map<ID, FEE> for an extended fees.
chain_parameters has a vector<parameter_extension> for extra parameters

We can always add extra operations for setting future chain_parameters, but it seems like extending that operation makes the most sense.   

Other operations may want to have extended data in the future, but at the moment our only option is to define a second version of the operation.   

What general philosophy will we have toward updating the protocol in the future?

**Comment by theoreticalbts** [comment]
This was a discussion we had some months ago.  My understanding was that the discussion concluded that adding extension points to the protocol was a non-goal; extensions should always be hardforks.
- On the one hand, since I was on the losing side of that debate, I am moderately pleased that this ticket seems to add momentum in the direction I've been pushing all along.
- On the other hand, I am worried that this ticket's assignment to alpha milestone is not correct; it should be moved to a later milestone, perhaps even after full release.

**Comment by vikramrajkumar** [comment]
Anything that goes into BitShares 2.0 that requires hardforking current Graphene needs to be done before Alpha.

If this is a non-goal for BitShares 2.0 as you said, or it does not require any hardforks, then we can defer it.

**Comment by nathanielhourt** [comment]
I don't think the purpose of this is to avoid hard-forking to add new features, which seems to be what @theoreticalbts is talking about. The purpose of this is to allow future versions of the code, which support new features, to process old blocks without special casing on fork blocks like BitShares 0.x does.

**Comment by theoreticalbts** [comment]
So I guess you want a `change_chain_semantics` op which contains a blob.  If you don't know how to parse the blob, the op is invalid.  If you do know how to parse the blob, it changes the chain semantics in some way based on parameters decoded from the blob.

The point of having the blob would be to leave the problem of deciding how to parse the blob until later, when you know what you want the hardfork to do.

This would allow you to change the semantics without hardcoding a fork block -- the fork block would be the block where the `change_chain_semantics` op was processed.

**Comment by theoreticalbts** [comment]
One problem is if we want a semantics change which changes operation serialization.

I think we can get away with this if we create a global, on-chain _operation meta table_ which contains, for each operation ID, an ordered list of field names and types, which can be used to generate a deserializer which deserializes old operation formats into new operation data structures (provided that default initialization of the new field(s) in the new operation data structure results in equivalent semantics to the old operation).

Processing a general operation meta table would require something like `field_reflector`, and may have negative performance implications.  Fortunately, we can sidestep the issue by automatically pre-generating at compile time serializers for all operation meta tables corresponding to past and planned future hardforks.  This would make meta-tables for old operation formats permanently part of the source code.  However, it would still greatly improve the situation from BTS 0.x:  The changes would not rely on hard-coded block numbers, and the code to deal with meta handling would be mostly confined to deserialization.

Deserialization would have to be substantially re-architected since it would now depend on chain state.

**Comment by arhag** [comment]
> The purpose of this is to allow future versions of the code, which support new features, to process old blocks without special casing on fork blocks like BitShares 0.x does.

The possible changes from hard forks can be so much bigger and incompatible than that though.

I think what is more generally important is for the code of each release to only need special cases to distinguish between _at most_ two version (pre hard fork and post hard fork). Each hard fork release would have two variants of the client and there would also be a database upgrade tool. 

The code of the first variant of the client code would need to support two versions of the blockchain protocol:  the current one (call this version V) and the next one (V+1) that it might instantaneously hard fork to. It would also only support version V' of the database state which in general is different than version V of the database state and needs to be designed so that it can support database state changes for both before the hard fork and after the hard fork (though it doesn't need to be as optimized as a post hard fork database could be). 

The second variant of the version (V+1) client would only support version (V+1) of the blockchain protocol and version (V+1) of the database state, which is optimized to only support the version (V+1) blockchain protocol. The second version of the client cannot be run until the hard fork has already happened and there is no chance of a chain reorganization unwinding that. 

A version V' snapshot of the database state can be generated using the upgrade tool using a version V snapshot of the database state. The upgrade tool is also able to convert a version V' snapshot of database state into a version (V+1) snapshot of the database state.

The idea is to first smoothly upgrade nodes from old clients (the second variant of version V client) to the first variant of the version (V+1) client. First, one copy of the old client (not a block producing one if the node is a witness) is shut down to get the serialized database state committed to disk. The upgrade tool runs on this database state (version V) to produce a new version V' database state snapshot. Then at least two copies of the first variant of the version (V+1) client run on that new database state snapshot. If the node is a witness, they smoothly transition by disabling block generation on the old client and then enabling it on one of the copies of the first variant of the version (V+1) client.

Then they wait until the on-chain condition is met that triggers the hard fork, which would automatically switch over the blockchain protocol to version (V+1) in the next maintenance period. Then wait some more if necessary to prevent any possible reorganizations. 

Finally, smoothly upgrade nodes from the first variant of the version (V+1) client to the second variant of the version (V+1) client. First, one copy (the one not producing blocks if the node is a witness) of the first variant of the version (V+1) client is shut down to get the serialized database state (version V') committed to disk. The upgrade tool runs on this database state to produce a new version (V+1) database state snapshot. Then at least two copies of the second variant of the version (V+1) client run on that new database snapshot state. If the node is a witness, they smoothly transition by disabling block generation on the remaining first variant of the version (V+1) client that is still running and enabling it on one of the copies of the second variant of the version (V+1) client. Now the newest clients are running on blockchain protocol and database state version (V+1) which is the only protocol and database version supported in the code used to compile these clients (no code bloat with conditionals for previous versions). 

Upgrading from version V to (V+K) for K > 1 all at once becomes more of a nuisance for users (doable of course but it requires K separate compilations and the above process repeated K times), but this should be an unlikely scenario since hard forks should be very rare anyway and most full node users should be on top of things. Anyway, in that case the user would likely just use the latest (second variant) client and a trusted snapshot of a database state (already at version V+K) as of a block after the hard fork from blockchain protocol version V+K-1 to V+K had already occurred.

**Comment by bytemaster** [comment]
The purpose of this ticket was only to enable easy extension of the SERIALIZED data.  When it comes to deserializing a structure it is VERY DIFFICULT (near impossible) to condition it based upon consensus state.     

I only want the ability to append data to existing data structures such that when we update the code we can read both OLD and NEW versions with the new code without having to resort to block number checks.   OLD code will not be able to read the NEW data.    

Ultimately this involves adding a flat_set<static_variant<void_t> > to the end of every structure we may want to extend in the future.  Each "extension" can be implemented by adding an extra type to the static_variant.   

It comes at the cost of one extra byte of serialization in the normal mode and a dynamic memory allocation if we ever have to extend the struct.

**Comment by bytemaster** [comment]
I have refactored fees via issue #144 to make them easily extensible at both a protocol and implementation level.
